project/AlgebraBoilerplate.scala

```diff
@@ -0,0 +1,162 @@
+/**
+ * Generate a range of boilerplate classes that would be tedious to write and maintain by hand.
+ *
+ * Copied, with some modifications, from
+ * [[https://github.com/milessabin/shapeless/blob/master/project/Boilerplate.scala Shapeless]].
+ *
+ * @author Miles Sabin
+ * @author Kevin Wright
+ */
+object AlgebraBoilerplate {
+  import scala.StringContext._
+
+  implicit class BlockHelper(val sc: StringContext) extends AnyVal {
+    def block(args: Any*): String = {
+      val interpolated = sc.standardInterpolator(treatEscapes, args)
+      val rawLines = interpolated.split('\n')
+      val trimmedLines = rawLines.map(_.dropWhile(_.isWhitespace))
+      trimmedLines.mkString("\n")
+    }
+  }
+
+  val templates: Seq[Template] = Seq(
+    GenTupleInstances
+  )
+
+  val header = "// auto-generated boilerplate"
+  val maxArity = 22
+
+  /**
+   * Return a sequence of the generated files.
+   *
+   * As a side-effect, it actually generates them...
+   */
+  def gen(dir: File): Seq[File] = templates.map { template =>
+    val tgtFile = template.filename(dir)
+    IO.write(tgtFile, template.body)
+    tgtFile
+  }
+
+  class TemplateVals(val arity: Int) {
+    val synTypes = (0 until arity).map(n => s"A$n")
+    val synVals = (0 until arity).map(n => s"a$n")
+    val `A..N` = synTypes.mkString(", ")
+    val `a..n` = synVals.mkString(", ")
+    val `_.._` = Seq.fill(arity)("_").mkString(", ")
+    val `(A..N)` = if (arity == 1) "Tuple1[A0]" else synTypes.mkString("(", ", ", ")")
+    val `(_.._)` = if (arity == 1) "Tuple1[_]" else Seq.fill(arity)("_").mkString("(", ", ", ")")
+    val `(a..n)` = if (arity == 1) "Tuple1(a)" else synVals.mkString("(", ", ", ")")
+  }
+
+  /**
+   * Blocks in the templates below use a custom interpolator, combined with post-processing to
+   * produce the body.
+   *
+   * - The contents of the `header` val is output first
+   * - Then the first block of lines beginning with '|'
+   * - Then the block of lines beginning with '-' is replicated once for each arity,
+   *   with the `templateVals` already pre-populated with relevant relevant vals for that arity
+   * - Then the last block of lines prefixed with '|'
+   *
+   * The block otherwise behaves as a standard interpolated string with regards to variable
+   * substitution.
+   */
+  trait Template {
+    def filename(root: File): File
+    def content(tv: TemplateVals): String
+    def range: IndexedSeq[Int] = 1 to maxArity
+    def body: String = {
+      val headerLines = header.split('\n')
+      val raw = range.map(n => content(new TemplateVals(n)).split('\n').filterNot(_.isEmpty))
+      val preBody = raw.head.takeWhile(_.startsWith("|")).map(_.tail)
+      val instances = raw.flatMap(_.filter(_.startsWith("-")).map(_.tail))
+      val postBody = raw.head.dropWhile(_.startsWith("|")).dropWhile(_.startsWith("-")).map(_.tail)
+      (headerLines ++ preBody ++ instances ++ postBody).mkString("\n")
+    }
+  }
+
+  object GenTupleInstances extends Template {
```

Comment:
This can be optimised to reduce the number of generated classes - see #3871
Can also be done post merge ofc

Reply:
I will wait to do this after #3871 is reviewed/merged in case there are any concerns. Essentially, it's trading static memory (generated classes) for dynamic memory (pointers to lambdas)?

Reply:
These days classes are also loaded in dynamic memory and can be unloaded by default.